eval/service.py

```diff
@@ -0,0 +1,695 @@
+
+
+async def run_agent_with_tracing(task: Task, browser: Browser | None = None, max_steps: int = 25):
+	try:
+		# Create task tracker
+		tracker = TaskTracker(task.task_id, task.confirmed_task)
+
+		browser = browser or Browser()
+		llm = ChatOpenAI(
+			model='gpt-4o',
+			temperature=0.0,
+		)
+
+		agent = Agent(task=task.confirmed_task, llm=llm, browser=browser)
+
+		# Pass our hook functions
+		result = await agent.run(max_steps=max_steps, on_step_start=tracker.on_step_start, on_step_end=tracker.on_step_end)
+
+		# Save final results
+		final_results = tracker.save_results()
+
+		return result
+	finally:
+		# Ensure proper cleanup
+		await asyncio.sleep(0.1)  # Give a moment for any pending tasks to complete
+		if not browser:
+			await agent.close()  # This will close the browser if we created it
+
+
+def judge_task_result(model, task_folder: Path, score_threshold: float = 3) -> Dict:
+	"""
+	Judge a single task result based on the success value of the final action.
+
+	Args:
+	    task_folder: Path to the task result folder
+
+	Returns:
+	    Dictionary containing judgment results
+	"""
+	result_file = task_folder / 'result.json'
+	if not result_file.exists():
+		return {'task_id': task_folder.name, 'judgement': None, 'success': False, 'error': 'No result.json found', 'score': 0.0}
+
+	try:
+		with open(result_file) as f:
+			result = json.load(f)
+
+		# If a Online_Mind2Web_evaluation is already saved, we can skip the eval
+		if result.get('Online_Mind2Web_evaluation'):
+			return result.get('Online_Mind2Web_evaluation')
+
+		# Get the screenshot paths, task description, and action history
+		screenshot_paths = result.get('screenshot_paths', [])
+		task_description = result.get('task')
+		action_history = result.get('action_history', [])
+
+		# Use the retry wrapper for evaluation
+		try:
+			eval_result = asyncio.run(
+				Online_Mind2Web_eval_with_retry(task_description, action_history, screenshot_paths, model, score_threshold)
+			)
+
+			if eval_result is None:
+				raise Exception('Evaluation failed after all retries')
+
+			messages, text, system_msg, record, key_points = eval_result
+
+			# Final steps to get judgement
+			judgement = model.invoke(messages).content
+
+			if 'success' in judgement.lower().split('status:')[1]:  # This is the official criteria for success
+				evaluation = {'task_id': task_folder.name, 'judgement': judgement, 'success': True, 'error': None, 'score': 1.0}
+			else:  # This is the official criteria for failure
+				evaluation = {'task_id': task_folder.name, 'judgement': judgement, 'success': False, 'error': None, 'score': 0.0}
+
+			# Save the Online_Mind2Web_evaluation into the result.json file
+			result['Online_Mind2Web_evaluation'] = evaluation
+			with open(result_file, 'w') as f:
+				json.dump(result, f, indent=2)
+
+			return evaluation
+
+		except Exception as e:
+			return {'task_id': task_folder.name, 'judgement': None, 'success': False, 'error': str(e), 'score': 0.0}
```

Comment:
hard to know what the actual error is if you stringify it without including the type, this is usually better because often the palin str msg alone is not descriptive enough:

```suggestion
			return {'task_id': task_folder.name, 'judgement': None, 'success': False, 'error': f'{type(err).__name__}: {err}', 'score': 0.0}
```